trin-core/src/portalnet/overlay_service.rs

```diff
@@ -297,26 +500,368 @@ impl OverlayService {
+    async fn connect_node(&mut self, node: Node, connection_direction: ConnectionDirection) {
+        let node_id = node.enr().node_id();
+        let key = kbucket::Key::from(node_id);
+        let status = NodeStatus {
+            state: ConnectionState::Connected,
+            direction: connection_direction,
+        };
+
+        let mut node_to_ping = None;
         match self
-            .discovery
-            .send_talk_req(
-                destination,
-                self.protocol.clone(),
-                Message::Request(request).to_bytes(),
-            )
+            .kbuckets
+            .write_with_warn()
+            .await
+            .insert_or_update(&key, node, status)
+        {
+            InsertResult::Inserted => {
+                // The node was inserted into the routing table. Add the node to the ping queue.
+                debug!(
+                    "[{:?}] New connected node added to routing table. Node: {}",
+                    self.protocol, node_id
+                );
+                self.peers_to_ping.insert(node_id);
+            }
+            InsertResult::Pending { disconnected } => {
+                // The disconnected node is the least-recently connected entry that is
+                // currently considered disconnected. This node should be pinged to check
+                // for connectivity.
+                node_to_ping = Some(disconnected);
+            }
+            InsertResult::StatusUpdated {
+                promoted_to_connected,
+            }
+            | InsertResult::Updated {
+                promoted_to_connected,
+            } => {
+                // The node existed in the routing table, and it was updated to connected.
+                if promoted_to_connected {
+                    debug!(
+                        "[{:?}] Node promoted to connected. Node: {}",
+                        self.protocol, node_id
+                    );
+                    self.peers_to_ping.insert(node_id);
+                }
+            }
+            InsertResult::ValueUpdated | InsertResult::UpdatedPending => {}
+            InsertResult::Failed(reason) => {
+                self.peers_to_ping.remove(&node_id);
+                debug!(
+                    "[{:?}] Could not insert node. Node: {}, Reason: {:?}",
+                    self.protocol, node_id, reason
+                );
+            }
+        }
+
+        // Ping node to check for connectivity. See comment above for reasoning.
+        if let Some(key) = node_to_ping {
+            match self.kbuckets.write_with_warn().await.entry(&key) {
+                kbucket::Entry::Present(ref mut entry, _) => {
+                    self.ping_node(&entry.value().enr()).await;
+                }
+                kbucket::Entry::Pending(ref mut entry, _) => {
+                    self.ping_node(&entry.value().enr()).await;
+                }
+                _ => {}
```

Comment:
Good question. After another look, I think we only want to ping the peer if the entry is `Entry::Present`. Otherwise, this node may have already been evicted by some other insertion of a more recently connected node.

My understanding is that the reason that we want to ping this node is because we prefer longer-lived peers. When we tried to perform the insert for the other node, the target bucket was full, so the node we tried to insert was inserted in a pending state. This `node_to_ping` is the least recently connected node in the target bucket whose connection state is now disconnected. We try to restore this `node_to_ping` to a connection state of connected with this ping. If the node responds, then that node will remain in the bucket and the node we inserted will remain in a pending state. Otherwise, the `node_to_ping` will eventually be evicted and replaced with the node we inserted.

This mechanic can be a bit difficult to explain, but hopefully that was helpful. There's more documentation [here](https://github.com/sigp/discv5/blob/master/src/kbucket.rs#L167). Nonetheless, good callout.